rfcs/0074-refine-s3-source-in-lazy-way.md

```diff
@@ -0,0 +1,87 @@
+The good news is that the barrier can assist in this process. When each checkpoint arrives, the source executor can 
+write the filenames of the finished files from the state cache into the barrier. 
+Eventually, through barrier collection, this information is transmitted to the Source Manager, completing the process. 
+The changes involved here include adding a new field or Mutation called `FinishedFiles` in the barrier 
+and introducing a new state `finishedFiles` in `Source Manager`. 
+This status will record the filenames of the finished or deleted files and needs to be persisted.
+Additionally, the `collect_barrier` needs to be modified to support retrieving `FinishedFiles` from the finished barriers.
+
+Another optimization is that since each barrier will carry at most one mutation, the content of Add/SplitChange 
+can be extracted when flowing through the `source executor`, reducing the overhead of subsequent transmissions.
+
+### Control the Number of Ongoing Files
+
+In the previously mentioned solution, we focused on how to avoid sending too many files during the reading process. 
+However, if we send all the files in the s3 bucket at once, we still face issues with an overly large barrier and prolonged list objects times. 
+Therefore, we can attempt to control the number of files sent each time, allowing us to manage the overall load on the cluster.
+
+From our introduction to the S3 Source, we know that the Source Manager can determine the total number of files 
+and the number of files that have been completed. With this information, we can ascertain the number of files currently being read, aka. ongoing files. 
+Taking into account the limitations of the S3 API, I propose a solution based on reading file list pagination 
+according to the number of files being read.
+
+1. Define the current page `cur_page`, initialized to 1 (indicating the max page per `list objects` request), 
+and the trigger to read the next page's file count `MIN_ONGOING_FILE_NUM`. This value should be less than 1000.
+
+2. During initialization, read all the files on the first page and add them to the reading queue. Assuming there's pagination,
+the number of ongoing files would be 1000.
+
+3. As the reading progresses, the number of files being read will gradually decrease. 
+When the number of ongoing files drops below `MIN_ONGOING_FILE_NUM`, allow the next page of the file list to be read; i.e., `cur_page += 1`.
+
+4. When the next enumerator tick arrives, read the file lists of pages `[0, cur_page)` and add the unread files in the pages to the reading queue.
+
+During this process, when the number of files being read exceeds `MIN_ONGOING_FILE_NUM`, 
+even if the request results on the first page change, the list of files being read remains unchanged. 
+The S3 API ensures that the listed file list is ordered, so we won't get significantly different results when listing objects consecutively.
```

Comment:
Can you elaborate more on "ordered"? Are they ordered by the time of creation or by file name? Is it possible that we miss some files by listing in a paginated way?

Reply:
> Objects are returned sorted in an ascending order of the respective key names in the list. [source](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html)

I think every time we call `list_objects`, AWS will take a snapshot of filenames and we use `continuation_token` to fetch the next page.